qiskit/transpiler/__init__.py

```diff
@@ -40,629 +41,787 @@
+    from qiskit.transpiler import PassManager
+    from qiskit.transpiler.passes import ALAPScheduleAnalysis, PadDynamicalDecoupling
+    from qiskit.transpiler.passes import CXCancellation, InverseCancellation
+
+    backend_durations = backend.target.durations()
+    dd_sequence = [XGate(), XGate()]
+    scheduling_pm = PassManager([
+        ALAPScheduleAnalysis(backend_durations),
+        PadDynamicalDecoupling(backend_durations, dd_sequence),
+    ])
+    inverse_gate_list = [
+        HGate(),
+        (RXGate(np.pi / 4), RXGate(-np.pi / 4)),
+        (PhaseGate(np.pi / 4), PhaseGate(-np.pi / 4)),
+        (TGate(), TdgGate()),
+
+    ])
+    logical_opt = PassManager([
+        CXCancellation(),
+        InverseCancellation([HGate(), (RXGate(np.pi / 4), RXGate(-np.pi / 4))
+    ])
+
+
+    # Add pre-layout stage to run extra logical optimization
+    pass_manager.pre_layout = logical_opt
+    # Set scheduling stage to custom pass manager
+    pass_manager.scheduling = scheduling_pm
+
+
+Then when :meth:`~.StagedPassManager.run` is called on ``pass_manager`` the
+``logical_opt`` :class:`~.PassManager` will be called prior to the ``layout`` stage
+and for the ``scheduling`` stage our custom :class:`~.PassManager`
+``scheduling_pm`` will be used.
+
+Custom Pass Managers
+====================
+
+While modifying the built-in preset pass managers to alter a particular stage
+of the compilation another common use case is to construct a custom pass
+manager to build an entirely custom pipeline for transforming the input
+circuits. You can leverage the :class:`~.StagedPassManager` class directly to do
+this. You can define arbitrary stage names and populate them a :class:`~.PassManager`
+instance. For example::
+
+    from qiskit.transpiler.passes import *
+    from qiskit.transpiler import PassManager, StagedPassManager
+
+    basis_gates = ["rx", "ry", "rxx"]
+    init = PassManager([UnitarySynthesis(basis_gates, min_qubits=3), Unroll3qOrMore()])
+    translate = PassManager(
+        [
+            Collect2qBlocks(),
+            ConsolidateBlocks(basis_gates=basis_gates),
+            UnitarySynthesis(basis_gates),
+        ]
+    )
+
+    staged_pm = StagedPassManager(
+        stages=["init", "translation"], init=init, translation=translate
+    )
+
+will create a new :class:`~.StagedPassManager` that has 2 stages ``init`` and ``translation``.
+There is no limit on the number of stages you can put in a custom :class:`~.StagedPassManager`
+instance.
+
+If you're building a custom pass manager the :ref:`stage_generator` functions may be useful
+for the construction of these pass managers. They are functions that provide common functionality
+which are used in most pass managers. For example :func:`~.generate_embed_passmanager` is used
+to "embed" a selected initial layout from a layout pass to the specified target device. Any
+layout pass that is working with a :class:`~.Layout` this will be needed to apply the layout.
```

Comment:
```suggestion
to "embed" a selected initial layout from a layout pass to the specified target device. For any
layout pass that is working with a :class:`~.Layout` this will be needed to apply the layout.
```